Initial version of (all but one) reporting org endpoints

simon-20 opened the pull request:
This PR:
Improves the test setup, and adds tests for:
- the two endpoints for fetching reporting orgs
- the endpoint which creates reporting orgs
- the endpoint which lists datasets

Improves some scaffolding.

Implements an initial version of the reporting-orgs end points.

Notes:
- The FGA code needs some custom exceptions adding
- A couple of the endpoints need additional app logging (auditing logging is done)
- The compensatory actions for rolling back when `create_reporting_org` fails are not complete

simon-20 committed:
test: artefact files for mocked SuiteCRM responses

simon-20 committed:
test: improvements to test framework

simon-20 committed:
test: reporting orgs list & create, list datasets

simon-20 committed:
feat: additions to fine grained auth code

simon-20 committed:
feat: improvements to scaffolding

simon-20 committed:
feat: get, create, update reporting org; get users

chrisarridge commented on tests/helpers/mocking.py:
```diff
@@ -203,11 +208,23 @@ async def test_lifespan(app: FastAPI) -> AsyncIterator[None]:
             subject=str(self._mocked_user_ids[test_user_num]),
             audience="iati_register_your_data",
-            scopes="ryd ryd:reporting_org",
+            scopes="ryd ryd:reporting_org ryd:reporting_org:create ryd:reporting_org_update ryd:dataset",
```
`ryd:reporting_org_update` -> `ryd:reporting_org:update`

chrisarridge commented on src/register_your_data_api/data_handling/data_schemas.py:
```diff
@@ -1,7 +1,69 @@
+    role: str
+
+
+class DatasetMetadata(pydantic.BaseModel):
```
Will need to add a human readable name to match the revised API spec.

simon-20 replied:
Yep, did you want me to add this to this PR?

chrisarridge replied:
I think it makes sense to do that here.

chrisarridge commented on src/register_your_data_api/routers/reporting_orgs.py:
```diff
@@ -69,123 +80,294 @@ def get_reporting_orgs(
-    # TODO: query SuiteCRM to get the list of reporting org actions
-    if include_actions == "yes":
-        pass
+    crm_reporting_orgs = crm.get_records("Accounts", page_number=1, page_size=10, fields=fields, filters=filters)
```
Why not use `crm.get_record_by_id` then you don't have to do any filtering and SuiteCRM doesn't have to do any filtering logic?

simon-20 replied:
I didn't use `get_record_by_id()` for the reason explained here: https://github.com/IATI/iati-registry-libsuitecrm/issues/7

simon-20 replied:
I can add issues to the two repositories to change this later, if you think that would be best solution?

chrisarridge replied:
Got it.  Thanks for explaining.  Let's leave it and we can decide later if we want to change it (e.g., with a change to libsuitecrm).

chrisarridge commented on src/register_your_data_api/routers/reporting_orgs.py:
```diff
@@ -69,123 +80,294 @@ def get_reporting_orgs(
+        # reporting org in SuiteCRM, but only if user is not operating in the
+        # capacity as a super_admin
+
+        # 3. Create a fine-grained authorisation for this user to be EDITOR of new reporting_org
```
Should be `ADMIN` (it is correct in code)

chrisarridge commented on src/register_your_data_api/routers/reporting_orgs.py:
```diff
@@ -69,123 +80,294 @@ def get_reporting_orgs(
+
+    crm.fetch_access_token()
+
+    users_for_org_from_suitecrm = crm.get_relationship("Accounts", str(org_id), "Contacts")
```
This should come from the FGA database not from SuiteCRM.  This will return all the relationships between Accounts and Contacts - including lots of relationships that have been added by the Secretariat and which have nothing to do with publishing.    We could add a separate relationship between accounts and contacts that just stores publishing, but that would mean that we were duplicating roles between the CRM and the FGA DB.

simon-20 replied:
Ah yes, good point, thank you. Will get that sorted.

simon-20 replied:
@chrisarridge, just to clarify: the iteration needs to happen on the list pulled from the FGA provider, but the list from SuiteCRM will still need to be pulled, right, because the name and email come from the CRM?

chrisarridge replied:
Yes!! Good point :)

chrisarridge reviewed (commented):
A few remarks but the real change/discussion point is in the code for getting the list of users - that needs to be sourced from the database, not from the CRM.

simon-20 committed:
fix: scope name corrected

simon-20 committed:
feat: fill all user details, from FGA & CRM

The user->org list from the FGA is the primary list
to iterate over in this context, because many user->
org relationships in CRM aren't to do with publishing,
but we pull in from CRM as needed.

Returns 500 if there is an inconsistency.

simon-20 committed:
fix: add missing field from dataset

simon-20 committed:
test: check content of dataset endpoint
